### DISCUSSION TITLE
Possibility to introduce `.clang-format` for AzerothCore project

### DISCUSSION THREAD

**Discussion by MichalVytrhlik** [azerothcore/azerothcore-wotlk#20750] [General]
Hey guys, I recently started to code for AC and I noticed there's a missing `.clang-format` file for formatting the project code. I also found out it was removed almost 3 years ago in this [pull request](https://github.com/azerothcore/azerothcore-wotlk/pull/9934). But I can't understand why it was removed, maybe I'm missing some information that I'm not aware of.
I think it would be very helpful to have `.clang-format`, so there's unified formatting for the project. It's at least easier for the IDE (I'm using CLion) to have the same formatting as the project wants.
What do you think?

**Comment by sogladev** [comment]
@gontrannopier  To answer your question
> But I can't understand why it was removed, maybe I'm missing some information that I'm not aware of.

The previous PR where `.clang.format` was included did not include any other changes to the project.
Anytime clang-format ran on a file, it applied the style rules to the entire file.
Furthermore, some editors detect `.clang.format` and automaticly run clang-format for you. 
So, you would get a bunch of changes that were not indended.

> .clang-format file for formatting 

There exists a `.editorconfig` in the root project. This file primarily focuses on basic editor settings rather than cpp specific formatting. That partly solves the "unified formatting" 

## Current standards

Acore uses a [wiki article that contains cpp code standards](https://www.azerothcore.org/wiki/cpp-code-standards)  together with CI python script to enforce some rules. I've found that it is not enough as sometimes a manual review still fails due to style. This should be avoided.

An option is to extend the wiki or add more rules to the codestyle-check CI, but clang-format should be **the** tool for this job.

Clang-format does come with a large caviat if I read correctly: You're not able to pick and choose which rules are enforced. It's all or nothing.

AzerothCore is quite lenient when it comes to codestyle. There are only a few rules compared to what clang-format enforces. I've found that applying a style brings massive changes, well over 100K+ lines changed.

## Writing a config
clang-format docs: https://clang.llvm.org/docs/ClangFormatStyleOptions.html
try out format: https://clang-format-configurator.site/
modified sample: https://gist.github.com/sogladev/a72d69ebab086d14e5ebf7a9feb57279
clang-format for 1 file: `clang-format -i somefile.cpp`

I've researched, wrote and applied a config for clang-format v18.1.3

I used the above tool with a modified sample to create a config that matches acore's style. It's not perfect and has these issues:
1. clang-format cannot differentiate between aligning enums and variables: https://github.com/llvm/llvm-project/issues/52983
2. breaking braces when defining a struct or a static is not possible https://github.com/llvm/llvm-project/issues/79054
3. some options may be set incorrectly that mismatch acore's style
4. personal opinion for some

below is a diff of acore with `.clang-format` and clang-format applied to all `.cpp` files in `/src/server/` :
https://github.com/azerothcore/azerothcore-wotlk/compare/master...sogladev:azerothcore-wotlk:clang-format-1?expand=1

## Conclusion
To me it's feasible to include a `.clang.format`, but it needs to be done carefully. The first step is to decide on a config that works. This is opinionated as different developers have different preferences.

I'm curious other people think about this.

**Reply by blinkysc** [reply]
This needs to be implemented lol can really simplify code-style

**Comment by ro-kue** [comment]
Just a little hint: You can exclude commits from the git history. I've done format changes a few times now in larger projects and the modus operandi is always:

1. Come to an agreement on style and linting for all team members
2. Write linting and style rules in appropriate format (clang-format and clang-tidy in this case)
3. Commit (So you can have a commit message for the rule you added)
4. auto fix whole project
5. fix manual problems that can't be auto fixed or write follow up issues for specific checks you disable because it is simply too much work to just do it on the side
6. Commit
7. Add commit hash of previous commit to .git-blame-ignore-revs file
8. Commit 

Now the styling changes don't show up in the commit history or blame. Every new PR needs to be rebased to master to not cause CI issues.

This worked very cleanly the 2 or 3 times I did that.
